### DISCUSSION TITLE
Opcode 17 error

### DISCUSSION THREAD

**Discussion by Yzgll** [sarchlab/mgpusim#60] [Q&A]
Mr. Sun, I have tried all versions of ROCm but have not been able to resolve opcode 17 error. I even created a new AES128 benchmark but still ended up with an opcode 17 error. I have also checked the alu_flat.go file, which contains opcode 16, 18, 20, 21, 23, 28, 29, 30 31. Do I need to implement opcoed 17 by myself or is it caused by other problems? Looking forward to your reply!

**Comment by syifan** [comment]
Yes. If the compiler is generating an instruction that is not supported by MGPUSim, you do need to implement it. You can first disassemble the code to make sure what is the instruction. I guess it is a `flat_load_sbyte`?  Then you can implement it. The file `alu_flat.go` is for emulation. For the instruction, you may also need to implement the timing part. Probably you can check this function that determines how to handle vector data return from memory https://github.com/sarchlab/mgpusim/blob/v3/timing/cu/computeunit.go#L504.

**Reply by syifan** [reply]
Would you be able to create a PR for us?

**Reply by Yzgll** [reply]
Thank you for your reply, I am working on it. I added two functions to solve opcode 17 and opcode2, and the program ended up running successfully, but I verified it and found Mismatch error. I think it is because the implementation logic of the two functions I added is incorrect. Since I hardly knew anything about disassembly, I didn't know the specific information of the instruction, so the function I added did not implement the correct function, which caused a mismatch. Could you please tell me how to get specific information about unimplemented instructions? Here are the two functions I added.
![1](https://github.com/sarchlab/mgpusim/assets/165646862/b226bd10-4924-46e6-8351-97c03eb871d5)
![2](https://github.com/sarchlab/mgpusim/assets/165646862/8f7503f8-2877-4d12-a3c7-4e360f53a821)

**Reply by syifan** [reply]
You can find AMD ISA documentation here. 

https://www.amd.com/system/files/TechDocs/gcn3-instruction-set-architecture.pdf

Chapter 12/13 has concise lists of the instructions. Note that there are some problems with the flat load instructions in the documentation.

**Reply by syifan** [reply]
Also, you should consider writing unit tests for these instructions. You can check the test files in the same folder.

**Reply by Yzgll** [reply]
With your help, I added some instructions and tests today and successfully solved opcode error, which finally ran successfully and passed the verification. But as you said earlier, I didn't implement the timing part, so it says "slice bounds out of range." . After that, I have modified the `computeunit.go` file `handleVectorDataLoadReturn` function, inside added opcode 17, as shown in the figure below. It was strange that the timing part did not raise the error, but it could not pass the verification, so I had to restore the modified place. I don't know how to modify computeunit, I hope you can give me some suggestions.
![3](https://github.com/sarchlab/mgpusim/assets/165646862/ae3ff7ee-3bcc-4611-854a-10525fbbf4f3)
